src/wp-includes/formatting.php

```diff
@@ -3242,89 +3242,47 @@ static function( $matches ) {
-	$html_parts  = preg_split( $script_and_style_regex, $text );
-
-	foreach ( $html_parts as &$part ) {
-		$part = preg_replace_callback( '|<a\s([^>]*target\s*=[^>]*)>|i', 'wp_targeted_link_rel_callback', $part );
-	}
-
-	$text = '';
-	for ( $i = 0; $i < count( $html_parts ); $i++ ) {
-		$text .= $html_parts[ $i ];
-		if ( isset( $extra_parts[ $i ] ) ) {
-			$text .= $extra_parts[ $i ];
+	$p = new WP_HTML_Tag_Processor( $text );
+	while ( $p->next_tag( 'a' ) ) {
+		$target = $p->get_attribute( 'target' );
+		if ( null === $target ) {
+			continue;
 		}
-	}
-
-	return $text;
-}
-
-/**
- * Callback to add `rel="noopener"` string to HTML A element.
- *
- * Will not duplicate an existing 'noopener' value to avoid invalidating the HTML.
- *
- * @since 5.1.0
- * @since 5.6.0 Removed 'noreferrer' relationship.
- *
- * @param array $matches Single match.
- * @return string HTML A Element with `rel="noopener"` in addition to any existing values.
- */
-function wp_targeted_link_rel_callback( $matches ) {
-	$link_html          = $matches[1];
-	$original_link_html = $link_html;
-
-	// Consider the HTML escaped if there are no unescaped quotes.
-	$is_escaped = ! preg_match( '/(^|[^\\\\])[\'"]/', $link_html );
-	if ( $is_escaped ) {
-		// Replace only the quotes so that they are parsable by wp_kses_hair(), leave the rest as is.
-		$link_html = preg_replace( '/\\\\([\'"])/', '$1', $link_html );
-	}
-
-	$atts = wp_kses_hair( $link_html, wp_allowed_protocols() );
-
-	/**
-	 * Filters the rel values that are added to links with `target` attribute.
-	 *
-	 * @since 5.1.0
-	 *
-	 * @param string $rel       The rel values.
-	 * @param string $link_html The matched content of the link tag including all HTML attributes.
-	 */
-	$rel = apply_filters( 'wp_targeted_link_rel', 'noopener', $link_html );
 
-	// Return early if no rel values to be added or if no actual target attribute.
-	if ( ! $rel || ! isset( $atts['target'] ) ) {
-		return "<a $original_link_html>";
-	}
-
-	if ( isset( $atts['rel'] ) ) {
-		$all_parts = preg_split( '/\s/', "{$atts['rel']['value']} $rel", -1, PREG_SPLIT_NO_EMPTY );
-		$rel       = implode( ' ', array_unique( $all_parts ) );
-	}
+		$rel = $p->get_attribute( 'rel' );
+		$rel = true === $rel ? "" : $rel;
+		$link_text = "rel=\"{$rel}\"";
 
-	$atts['rel']['whole'] = 'rel="' . esc_attr( $rel ) . '"';
-	$link_html            = implode( ' ', array_column( $atts, 'whole' ) );
+		/**
+		 * Filters the rel values that are added to links with `target` attribute.
+		 *
+		 * @since 5.1.0
+		 *
+		 * @param string $rel       The rel values.
+		 * @param string $link_html The matched content of the link tag including all HTML attributes.
+		 */
+		$updated_rel = apply_filters( 'wp_targeted_link_rel', 'noopener', $link_text );
+		if ( ! $updated_rel ) {
+			continue;
+		}
 
-	if ( $is_escaped ) {
-		$link_html = preg_replace( '/[\'"]/', '\\\\$0', $link_html );
+		$all_parts = preg_split( '/\s/', "$rel $updated_rel", -1, PREG_SPLIT_NO_EMPTY );
```

Comment:
Isn't it the same logic as with class names? I wonder how many HTML attributes are like that – perhaps there's space for a more generic "add_flag()" or so.

Reply:
had the same thought 😉